### PR TITLE
Introduce a new function to convert an application model to backbone model.

### PR DESCRIPTION
**Pull request by qlzh727**
# What does this PR do?
WIP, this PR is more for a demo and not finalized yet.

The current PR will attach the output on each feature level to the model, so that we could receate the application model to a backbone model, so that features on each level could be consumed by decoder like FPN or RPN.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [X] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

### CODE REVIEW

**Comment by LukeWood** [comment]
Hey Scott are you looking for feedback on the approach, the API surface, or both?

I really like the approach - maybe we could promote backbone outputs to a public API even.

API wise maybe we name the function models.as_backbone(model) or maybe we make a BaseCvModel class.  If we go with the latter there are some other features we could add.  One that comes to mind is automatically unpacking dictionary inputs in the format {images, labels}.  This would allow us to drop the code to unpackage tf dataset dictionaries all over

File: keras_cv/models/resnet_v2.py [hunk 1]
```diff
@@ -355,10 +374,21 @@ def ResNetV2(
...
     if weights is not None:
         model.load_weights(weights)
+    # Set this private attribute for recreate backbone model with outputs at each of the
+    # resolution level.
+    model._backbone_level_outputs = stack_level_outputs
 
     return model
 
 
+def create_backbone_model(model):
```
Comment (tanzhenyu) [inline keras_cv/models/resnet_v2.py:384]: maybe add some documentation here, how we imagine users to use this?

**Comment by qlzh727** [comment]
> 

Both. The ideal API interface to me is like:

```
application_model = resnet_v2.Resnet50(....)
backbone_model = application_model.to_backbone_model()
```

In the application model building logic, it will create a new method `to_backbone_model` for the model, and populate the logic.

**Comment by LukeWood** [comment]
> Both. The ideal API interface to me is like:
> 
> ```
> application_model = resnet_v2.Resnet50(....)
> backbone_model = application_model.to_backbone_model()
> ```
> 
> In the application model building logic, it will create a new method `to_backbone_model` for the model, and populate the logic.

Yep, I agree this is the ideal interface.

what do you think about a CVModel base class that offers this method (raises error if no backbone outputs are set), and unpackages images/labels from the dict format in fit() for us?

File: keras_cv/models/resnet_v2.py (hunk 1 above)
Comment (qlzh727) [reply keras_cv/models/resnet_v2.py]: Done. Changed to a new interface.

**Comment by ianstenbit** [comment]
> what do you think about a CVModel base class that offers this method (raises error if no backbone outputs are set), and unpackages images/labels from the dict format in fit() for us?

I personally like this idea, but it's also a bit more frameworky than just building a raw Keras model. One benefit is that the structure could be easily extended for e.g. DiffusionModels, which will likely share a lot of functionality and likely a custom train_step.

FWIW, I personally like `as_backbone` more than `to_backbone_model` for the API

**Comment by qlzh727** [comment]
> 

The new base class approach might be more complicated, since we need to handle all 3 models types (Functional, subclassing, Sequential). We could have a try for this approach, but I feel this is bit orthogonal to the current change.

**Review by LukeWood — APPROVED** [review]
I agree that the class approach is out of scope for this.  All in all, looks great and I really like this API.

File: keras_cv/models/resnet_v2.py [hunk 2]
```diff
@@ -355,6 +376,37 @@ def ResNetV2(
...
+    # Set this private attribute for recreate backbone model with outputs at each of the
+    # resolution level.
+    model._backbone_level_outputs = stack_level_outputs
+
+    def as_backbone(self):
+        """Convert the Resnet application model into a model backbone for other tasks.
+
+        The backbone model will usually take same inputs as the original application
+        model, but produce multiple outputs, one for each feature level. Those outputs
+        can be feed to network downstream, like FPN and PRN.
```
Comment (tanzhenyu) [inline keras_cv/models/resnet_v2.py:388]: I think you mean RPN?

**Review by tanzhenyu — APPROVED** [review]

File: keras_cv/models/resnet_v2.py [hunk 3]
```diff
@@ -355,6 +376,37 @@ def ResNetV2(
...
+
+    def as_backbone(self):
+        """Convert the Resnet application model into a model backbone for other tasks.
+
+        The backbone model will usually take same inputs as the original application
+        model, but produce multiple outputs, one for each feature level. Those outputs
+        can be feed to network downstream, like FPN and RPN.
+
+        The output of the backbone model will be a dict with int as key and tensor as
+        value. The int key represent the level of the feature output.
```
Comment (martin-gorner) [inline keras_cv/models/resnet_v2.py:391]: Please define "level"
From "Practical ML for CV": "A typical feature pyramid has five levels corresponding to scales P3, P4, P5, P6, P7 in the backbone. Scale Pn represents a feature map 2n times smaller in width and height than the input image."
What does "level" stand for in the docstring ?
